File: csharp-selenium-nunit-reqnroll/SauceDemo.UiTests/Configuration/TestSettings.cs
using SauceDemo.UiTests.Drivers;

namespace SauceDemo.UiTests.Configuration;

/// <summary>Bound from the "TestSettings" section of appsettings*.json, overridable via TestSettings__* env vars.</summary>
public sealed class TestSettings
{
    public string BaseUrl { get; init; } = "https://www.saucedemo.com/";

    public BrowserType Browser { get; init; } = BrowserType.Chrome;

    public bool Headless { get; init; }

    /// <summary>Selenium Grid / remote WebDriver endpoint. Null means launch a local browser.</summary>
    public string? RemoteUrl { get; init; }

    public int ExplicitWaitSeconds { get; init; } = 20;

    public int PageLoadTimeoutSeconds { get; init; } = 30;

    /// <summary>How long a scenario waits in <see cref="Users.UserPool.Acquire"/> for a free pooled user before failing.</summary>
    public int UserAcquireTimeoutSeconds { get; init; } = 120;
}

File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/config/TestSettings.java
package com.saucedemo.uitests.config;

import java.time.Duration;

/**
 * Bound from the {@code TestSettings} section of appsettings*.json, overridable by the plain env vars
 * documented in this framework's README.
 *
 * @param remoteUrl a Selenium Grid / remote WebDriver endpoint; null means launch a local browser
 */
public record TestSettings(
        String baseUrl,
        BrowserName browser,
        boolean headless,
        String remoteUrl,
        int explicitWaitSeconds,
        int pageLoadTimeoutSeconds) {

    public Duration explicitWait() {
        return Duration.ofSeconds(explicitWaitSeconds);
    }

    public Duration pageLoadTimeout() {
        return Duration.ofSeconds(pageLoadTimeoutSeconds);
    }

    public boolean hasRemoteUrl() {
        return remoteUrl != null && !remoteUrl.isBlank();
    }
}
